### PR TITLE
Add support for include_fields and drop_fields

### PR DESCRIPTION
**Pull request by monicasarbu**
This PR adds two configuration options: `include_fields` and `drop_fields`. 

There is a list of mandatory fields like: `@timestamp`, `beat`, `count`, `status` that cannot be removed and they are always available in the exported fields. With the two configuration options you can configure the additional fields that will be exported or not exported. By default, all fields are exported.

Here are the options:

- `include_fields=[]` -> exports only the mandatory fields
- `include_fields is not defined` (commented) -> export all fields (the default option)
- `drop_fields=[]` -> export all fields (the default option)
 
As fields, they accept a map like `proc` or a value like `proc.cpu.total_p`. 

In case there is a field to include that is not available in the event, then just ignore the field. 

When multiple filtering rules are defined, we start with a copy of the event and apply the filtering rules one by to the
event.

<pre>
   event -> filter rule 1 -> event 1 -> filter rule 2 -> event 2 ...
</pre>

   where event is the initial event, event1 is the event resulted after applying "filter rule1" and it's considered input for the "filter rule 2" and so on.

### CODE REVIEW

File: libbeat/common/event_test.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+		},
+	}
+
+	for _, test := range tests {
+		assert.True(t, deepEqual(test.Output, ConvertToGenericEvent(test.Input)))
+	}
+
+}
+
+// deepEqual compars two MapStr assuming they only have primitive types as values
```
Comment (tsg) [inline libbeat/common/event_test.go]: s/compars/compares/

**Comment by pickypg** [comment]
> If you configure one of these fields in the drop_fields, it will be ignored.

Instead of ignoring it, can it just fail? That way if the list ever grows in the future, then a user that upgrades won't be surprised to find their rules no longer apply as they did before.

**Comment by monicasarbu** [comment]
@pickypg The list of mandatory fields is defined in the code, so the user can configure only the extra fields that he wants to include or drop. I agree with you that the sentence is confusing, so I updated the description. Thanks for noticing.

File: libbeat/beat/beat.go [hunk 2]
```diff
@@ -84,6 +85,7 @@ type BeatConfig struct {
 	Output  map[string]*ucfg.Config
 	Logging logp.Logging
 	Shipper publisher.ShipperConfig
+	Filter  []filter.FilterConfig
```
Comment (ruflin) [inline libbeat/beat/beat.go]: I suggest we don't introduce a new top level config here as we are in the process to get rid of it. This leads to problems when we want for example run multiple beats as one binary. We should only have one common name space which is beat.

File: libbeat/filter/config.go [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+type DropFieldsConfig struct {
+	Fields []string `config:"fields"`
+}
+
+type IncludeFieldsConfig struct {
+	Fields []string `config:"fields"`
+}
+
+type FilterConfig struct {
+	DropFields    *DropFieldsConfig    `config:"drop_fields"`
```
Comment (ruflin) [inline libbeat/filter/config.go]: I would suggest that we call it `exclude_fields` instead of `drop_fields` as we already have `exclude_lines`.

File: libbeat/beat/beat.go [hunk 4]
```diff
@@ -149,6 +151,7 @@ func (b *Beat) Start() error {
 	// Additional command line args are used to overwrite config options
 	err, exit := b.CommandLineSetup()
 	if err != nil {
+		logp.Err("fails to load command line setup")
```
Comment (ruflin) [inline libbeat/beat/beat.go]: No error were written here to the log as the errors are returned. It Should be part of the implementing beat to decide if the error should be written to log or not.

File: libbeat/beat/beat.go [hunk 5]
```diff
@@ -159,12 +162,14 @@ func (b *Beat) Start() error {
 	// Loads base config
 	err = b.LoadConfig()
 	if err != nil {
+		logp.Err("fails to load the config")
```
Comment (ruflin) [inline libbeat/beat/beat.go]: See above

File: libbeat/beat/beat.go [hunk 6]
```diff
@@ -159,12 +162,14 @@ func (b *Beat) Start() error {
...
 		return err
 	}
 
 	// Configures beat
 	err = b.BT.Config(b)
 	if err != nil {
+		logp.Err("fails to load the beat config")
```
Comment (ruflin) [inline libbeat/beat/beat.go]: See above

File: libbeat/beat/beat.go [hunk 7]
```diff
@@ -229,13 +234,20 @@ func (b *Beat) LoadConfig() error {
 
 	pub, err := publisher.New(b.Name, b.Config.Output, b.Config.Shipper)
 	if err != nil {
-		return fmt.Errorf("error Initialising publisher: %v\n", err)
+		return fmt.Errorf("error initializing publisher: %v\n", err)
+	}
+
+	filters, err := filter.New(b.Config.Filter)
+	if err != nil {
+		return fmt.Errorf("error initializing filters: %v\n", err)
```
Comment (ruflin) [inline libbeat/beat/beat.go]: We could apply the same to the errors above in case they are not good enough, means we "enhance" the error message.

File: libbeat/common/event.go [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+			case reflect.Float32, reflect.Float64:
+			case reflect.Complex64, reflect.Complex128:
+			case reflect.String:
+			case reflect.UnsafePointer:
+			case reflect.Array, reflect.Slice:
+			//case reflect.Chan:
+			//case reflect.Func:
+			//case reflect.Interface:
+			case reflect.Map:
+				anothermap, err := MarshallUnmarshall(value)
```
Comment (ruflin) [inline libbeat/common/event.go]: I remember we had some issue in the past with JSON marshalling / unmarshalling. Could this have quite an impact on performance?

File: libbeat/common/mapstr_test.go [hunk 9]
```diff
@@ -43,6 +43,126 @@ func TestMapStrUnion(t *testing.T) {
...
+				"c32": 2,
+			},
+		},
+	}
+
+	err := m.Delete("c.c2")
+	assert.Equal(nil, err)
+	assert.Equal(MapStr{"c": MapStr{"c1": 1, "c3": MapStr{"c31": 1, "c32": 2}}}, m)
+
+	err = m.Delete("c.c2.c21")
```
Comment (ruflin) [inline libbeat/common/mapstr_test.go]: This removes an non existent field. So it is expected that no error is returned here?

File: libbeat/common/event.go (hunk 8 above)
Comment (monicasarbu) [inline libbeat/common/event.go]: Yes, the plan is to call unmarshall/Marshall only when it is really needed. It shouldn't be called for official supported beats, but it might be needed for the community beats that are exporting the data into a structure instead of a Mapstr.

File: libbeat/beat/beat.go (hunk 2 above)
Comment (monicasarbu) [inline libbeat/beat/beat.go]: We can also do this change in a separate PR once it's merged.

Comment (ruflin) [inline libbeat/beat/beat.go]: I think if we merge it like that we probably also going to keep it that way as also tests below and config changes for topbeat were already made for it. I suggest to currently add it under `shipper` which should be `beat`, same as we have the fields etc.

Comment (monicasarbu) [inline libbeat/beat/beat.go]: I think `shipper` or `beat` (to be honest I think beat is too generic, maybe host?) should be the section that includes only the configuration options for that host like geoip configuration, max_queue, tags and shouldn't include `fields` and `filter` as these are not a configuration of the host and more of an enhancement of the output data. 
I suggest to leave it like this for now and change it in a separate PR if we decide t o structure differently the configuration file.

File: libbeat/filter/config.go (hunk 3 above)
Comment (monicasarbu) [inline libbeat/filter/config.go]: For me `exclude_lines` is a way to skip some log lines when reading the log file (before even creating the event), while `drop_fields` is something that you would do after having the event as an "enhancement" of the event. I think it's important to have different naming "drop" vs "exclude" to make the difference between when the filtering is done: before creating the event vs after creating the event.
In addition I think `drop_fields` goes well with `drop_event`. If we would rename `drop_fields` to `exclude_fields` then we also have to rename `drop_event` with `exclude_event` which sounds a bit strange to me.

File: libbeat/beat/beat.go (hunk 4 above)
Comment (monicasarbu) [inline libbeat/beat/beat.go]: These error messages somehow remain after rebasing with master. I am fixing them manually. Thanks for noticing it.

File: libbeat/common/mapstr_test.go (hunk 9 above)
Comment (monicasarbu) [inline libbeat/common/mapstr_test.go]: Fixed, thanks for noticing it!

File: libbeat/beat/beat.go (hunk 2 above)
Comment (ruflin) [inline libbeat/beat/beat.go]: https://github.com/elastic/beats/pull/1165 will allow to place it under beat.filter as we do we fields and tags.

File: packetbeat/tests/system/test_0060_filtering.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+        assert objs[1]["status"] == "OK"
+        assert objs[2]["status"] == "Error"
+
+        print objs[0]
+
+        assert "http.request_headers" in objs[0]
+        assert "http.response_headers" in objs[0]
+
+        # generic filtering fails to delete the htp.response_headers.transfer-encoding
+        assert "transfer-encoding" not in objs[0]["http.response_headers"]
```
Comment (tsg) [inline packetbeat/tests/system/test_0060_filtering.py]: Not sure I understand the comment above this statement. Does it really fail or it's more like "check if filtering fails to delete.."?

File: packetbeat/tests/system/test_0060_filtering.py [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+        self.run_packetbeat(pcap="http_minitwit.pcap",
+                            debug_selectors=["http", "httpdetailed"])
+        objs = self.read_output()
+
+        assert len(objs) == 3
+        assert all([o["type"] == "http" for o in objs])
+
+        assert objs[0]["status"] == "OK"
+        assert objs[1]["status"] == "OK"
+        assert objs[2]["status"] == "Error"
```
Comment (tsg) [inline packetbeat/tests/system/test_0060_filtering.py]: would be good to add some `include_fields` tests. Right now I only see all in or all out tests for `include_fields`.

Comment (tsg) [inline packetbeat/tests/system/test_0060_filtering.py]: Nevermind, I saw the topbeat tests below :-).

File: topbeat/tests/system/test_filtering.py [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+            "proc.mem.rss",
+            "proc.mem.rss_p"
+        ]:
+            assert key in output
+
+        for key in [
+            "proc.name",
+            "proc.pid",
+        ]:
+            assert key not in output
```
Comment (tsg) [inline topbeat/tests/system/test_filtering.py]: Maybe add tests to delete multiple full keys, like `drop_fields=["proc.mem.rss", "proc.cpu.total", "proc.cpu.user"]` and then another one with `include_fields` and the same list.

File: libbeat/common/mapstr.go [hunk 13]
```diff
@@ -57,6 +58,110 @@ func (m MapStr) Update(d MapStr) {
...
+			}
+		} else {
+			return fmt.Errorf("unknown key %s", keyPart)
+		}
+	}
+	delete(mapp, keyParts[keysLen-1])
+	return nil
+}
+
+func (m MapStr) CopyTo(to MapStr, key string) error {
```
Comment (tsg) [inline libbeat/common/mapstr.go]: Maybe a better name for this function would be `CopySubtree`? Alternatively a godoc comment about what it does would be good.

**Comment by tsg** [comment]
I think mandatory should really only be `@timestamp` and `type` because we rely on them in a lot of places. Everything else should be deletable.

**Comment by ruflin** [comment]
@tsg Agree with that these should be the required libbeat fields. We could add a possibility in the future that beats can also reserve some fields as required.

File: libbeat/filter/config.go (hunk 3 above)
Comment (ruflin) [inline libbeat/filter/config.go]: Makes sense. Would it be perhaps `SelectFields` instead of `IncludFields` if we apply the same logic?

File: libbeat/common/mapstr.go (hunk 13 above)
Comment (tsg) [inline libbeat/common/mapstr.go]: Or `CopyFieldsTo`.

**Comment by tsg** [comment]
LGTM apart of the comments above. Nice work!

Longer term we should work towards an Event type that's not `MapStr{}` so that all this can be safer and faster.

File: libbeat/filter/config.go (hunk 3 above)
Comment (monicasarbu) [inline libbeat/filter/config.go]: While thinking how to restructure the config file, I came up with the idea to rename `include_fields` with `drop_all_fields_except`. The name is rather long, but it describes exactly what it's designed for. I can do this change in a separate PR.

File: packetbeat/tests/system/test_0060_filtering.py (hunk 10 above)
Comment (monicasarbu) [inline packetbeat/tests/system/test_0060_filtering.py]: yes, exactly

File: libbeat/filter/config.go (hunk 3 above)
Comment (ruflin) [inline libbeat/filter/config.go]: I would prefer `select_fields` over `drop_all_fields_except` as it is shorter and I think it also describes the issue.

In any case I would do the renaming directly in the same PR.

**Comment by monicasarbu** [comment]
All the comments were addressed. It's ready for review.

**Comment by monicasarbu** [comment]
I changed the action from `include_fields` to `drop_all_fields_except` to be more easier to understand.

**Comment by ruflin** [comment]
@elastic/beats Any other thoughts on the naming of `include_fields`? https://github.com/elastic/beats/pull/1120#discussion_r56330744

**Comment by monicasarbu** [comment]
I renamed drop_all_fields_except with include_fields. Now it's ready for the final review.

**Comment by monicasarbu** [comment]
Rebased to master, it should be green soon.

**Comment by monicasarbu** [comment]
Finally all green now
